src/gg.regression.unity.bots/Runtime/Scripts/ActionManager/Actions/UISliderReleaseAction.cs

```diff
@@ -0,0 +1,166 @@
+
+    public class UISliderReleaseInstance : RGGameActionInstance<UISliderReleaseAction, float>
+    {
+        public UISliderReleaseInstance(UISliderReleaseAction action, Object targetObject) : base(action, targetObject)
+        {
+        }
+        
+        public static Image FindSliderBackground(Selectable slider)
+        {
+            return slider.gameObject.transform.GetComponentInChildren<Image>();
+        }
+
+        public static Vector2? GetMousePosForParam(float param, Selectable slider)
+        {
+            Image sliderBg = FindSliderBackground(slider);
+            if (sliderBg == null)
+                return null;
+            var ssBounds = RGActionManagerUtils.GetUIScreenSpaceBounds(sliderBg.gameObject);
+            if (!ssBounds.HasValue)
+                return null;
+            Vector2 min = ssBounds.Value.min;
+            Vector2 max = ssBounds.Value.max;
+            Vector2 center = ssBounds.Value.center;
+
+            if (slider is Slider uiSlider)
+            {
+                switch (uiSlider.direction)
+                {
+                    case Slider.Direction.LeftToRight:
+                        return new Vector2(Mathf.Lerp(min.x, max.x, param), center.y);
+                    case Slider.Direction.RightToLeft:
+                        return new Vector2(Mathf.Lerp(max.x, min.x, param), center.y);
+                    case Slider.Direction.BottomToTop:
+                        return new Vector2(center.x, Mathf.Lerp(min.y, max.y, param));
+                    case Slider.Direction.TopToBottom:
+                        return new Vector2(center.x, Mathf.Lerp(max.y, min.y, param));
+                    default:
+                        throw new Exception("Unexpected slider direction " + uiSlider.direction);
+                }
+            } else if (slider is Scrollbar uiScrollbar)
+            {
+                switch (uiScrollbar.direction)
+                {
+                    case Scrollbar.Direction.LeftToRight:
+                        return new Vector2(Mathf.Lerp(min.x, max.x, param), center.y);
+                    case Scrollbar.Direction.RightToLeft:
+                        return new Vector2(Mathf.Lerp(max.x, min.x, param), center.y);
+                    case Scrollbar.Direction.BottomToTop:
+                        return new Vector2(center.x, Mathf.Lerp(min.y, max.y, param));
+                    case Scrollbar.Direction.TopToBottom:
+                        return new Vector2(center.x, Mathf.Lerp(max.y, min.y, param));
+                    default:
+                        throw new Exception("Unexpected scrollbar direction " + uiScrollbar.direction);
+                }
+            }
+            else
+            {
+                return null;
+            }
+        }
+
+        protected override bool IsValidActionParameter(float param)
+        {
+            return true;
+        }
+
+        protected override IEnumerable<RGActionInput> GetActionInputs(float param)
```

Comment:
I guess one part I am confused about is exactly how the monkey bot decides where to click on a slider, or what to type into an input. Does the action encapsulate the whole "select slider AND click on a part of it", or is it just selecting, and a separate click action needs to happen? Same for the input fields - is the focus and key press part of that entire action, or are they all separate?

Reply:
Good question: you'll note that we have two actions for the slider (UISliderPress, UISliderRelease), and for some of the other UI types as well (e.g. InputField has separate actions for pressing/focusing it and text entry). This goes back to the fundamental design of the action manager, where every action is defined as a set of inputs applicable to a _single_ frame. 

For the slider, you'll note that both the press and release actions have a `float` parameter ranging from 0 to 1 that indicate where the mouse should be hovering over the slider. For the press action, it will move the mouse cursor to that location and press the mouse button down, and the release action will move the mouse cursor and release the mouse button. (I've added some comments to the code to elaborate on this)

For the input fields, there are separate actions for focusing the text field (this is generated as a UIObjectPressAction) and entering text (UIInputFieldTextEntryAction), since the latter has a precondition that the text field is already focused. You'll note that the text entry action checks that the input field is focused in its `IsValidForObject` function.

In general, it is up to the agent logic whether to prioritize doing certain sequences of actions (this could be a fixed heuristic, or something learned). As mentioned in the action manager [reference](https://github.com/Regression-Games/RGUnityBots/blob/main/src/gg.regression.unity.bots/Runtime/Scripts/ActionManager/README.md), the RGMonkeyBot has some heuristics to prioritize some common sequences of actions, e.g. releasing the mouse over the same location it was pressed with some greater chance.